rust/lancedb/src/embeddings/sentence_transformers.rs

```diff
@@ -0,0 +1,447 @@
+    /// Set the path to the tokenizer file.
+    /// Defaults to `tokenizer.json`
+    ///
+    /// Note: this is the path inside the huggingface repo, **NOT the path on disk**.
+    pub fn tokenizer_path<S: Into<String>>(mut self, tokenizer: S) -> Self {
+        self.tokenizer_path = Some(tokenizer.into());
+        self
+    }
+
+    /// Set the path inside the huggingface repo to the model file.
+    /// Defaults to `model.safetensors`
+    ///
+    /// Note: this is the path inside the huggingface repo, **NOT the path on disk**.
+    ///
+    /// Note: we currently only support a single model file.
+    pub fn model_path<S: Into<String>>(mut self, model: S) -> Self {
+        self.model_path = Some(model.into());
+        self
+    }
+
+    pub fn build(mut self) -> SentenceTransformersEmbeddings {
+        let model_id = self.model.as_deref().unwrap_or("all-MiniLM-L6-v2");
+        let model_id = format!("sentence-transformers/{}", model_id);
+        let config = self.config_path.as_deref().unwrap_or("config.json");
+        let tokenizer = self.tokenizer_path.as_deref().unwrap_or("tokenizer.json");
+        let model_path = self.model_path.as_deref().unwrap_or("model.safetensors");
+        let device = self.device.unwrap_or(Device::Cpu);
+
+        let repo = if let Some(revision) = self.revision {
+            Repo::with_revision(model_id, RepoType::Model, revision.to_string())
+        } else {
+            Repo::new(model_id, RepoType::Model)
+        };
+        let (config_filename, tokenizer_filename, weights_filename) = {
+            let api = Api::new().expect("failed to create api");
+            let api = api.repo(repo);
+            let config = api.get(config).expect("failed to get config");
+            let tokenizer = api.get(tokenizer).expect("failed to get tokenizer");
+            let weights = api.get(model_path).expect("failed to get weights");
+
+            (config, tokenizer, weights)
+        };
+
+        let config = std::fs::read_to_string(config_filename).expect("failed to read config");
+        let config: Config = serde_json::from_str(&config).expect("failed to parse config");
+
+        let mut tokenizer =
+            Tokenizer::from_file(tokenizer_filename).expect("failed to load tokenizer");
+        if self.padding.is_some() {
+            tokenizer.with_padding(self.padding.take());
+        }
+
+        let vb = unsafe {
+            VarBuilder::from_mmaped_safetensors(&[weights_filename], DTYPE, &device)
+                .expect("failed to load weights")
+        };
+        let model = BertModel::load(vb, &config).expect("failed to load model");
+        SentenceTransformersEmbeddings {
+            model,
+            tokenizer,
+            device,
+            n_dims: self.n_dims,
+        }
+    }
+}
+
+macro_rules! from_cpu_storage {
+    ($cpu_storage:ident, $data:ident, $builder:ident, $layout:ident, $embeddings:ident, $d1:ident, $d2:ident) => {{
+        let mut builder = $builder::new();
```

Comment:
I feel like you could pre-size the builder here.
```suggestion
        let mut builder = $builder::with_capacity(dim1 * dim2);
```